File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/device/DeviceEnvironment.java
package de.marmaro.krt.ffupdater.device;

import com.google.common.base.Preconditions;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

import static android.os.Build.SUPPORTED_ABIS;
import static android.os.Build.VERSION.SDK_INT;

/**
 * This class returns all supported ABIs and the API level.
 */
public class DeviceEnvironment {
    private final List<ABI> abiList;
    private final int sdkInt;

    public DeviceEnvironment() {
        this(findSupportedAbis(), SDK_INT);
    }

    public DeviceEnvironment(List<ABI> abiList, int sdkInt) {
        Preconditions.checkArgument(!abiList.isEmpty());
        this.abiList = abiList;
        this.sdkInt = sdkInt;
    }

    public int getApiLevel() {
        return sdkInt;
    }

    /**
     * An ordered list of ABIs supported by this device. The most preferred ABI is the first element in the list.
     *
     * @return a list of at least one element
     */
    public List<ABI> getSupportedAbis() {
        return abiList;
    }

    private static List<ABI> findSupportedAbis() {
        final HashMap<String, ABI> map = new HashMap<>();
        map.put("arm64-v8a", ABI.AARCH64);
        map.put("armeabi-v7a", ABI.ARM);
        map.put("x86_64", ABI.X86_64);
        map.put("x86", ABI.X86);

        return Arrays.stream(SUPPORTED_ABIS).filter(map::containsKey).map(map::get).collect(Collectors.toList());
    }
}

File: ffupdater/src/main/java/de/marmaro/krt/ffupdater/device/DeviceEnvironment.kt
package de.marmaro.krt.ffupdater.device

import android.os.Build
import android.os.Build.VERSION.SDK_INT

/**
 * This class returns all supported ABIs and the API level.
 * This indirection with abis and supportAndroidXXX() is necessary, because Mockk can't mock/change the
 * Android classes.
 */
object DeviceEnvironment {
    val abis = findSupportedAbis()
    val sdkInt = SDK_INT

    /**
     * API level 23
     */
    fun supportsAndroidMarshmallow(): Boolean {
        return SDK_INT >= Build.VERSION_CODES.M
    }

    /**
     * API level 24
     */
    fun supportsAndroidNougat(): Boolean {
        return SDK_INT >= Build.VERSION_CODES.N
    }

    /**
     * API level 26
     */
    fun supportsAndroidOreo(): Boolean {
        return SDK_INT >= Build.VERSION_CODES.O
    }

    /**
     * API level 28
     */
    fun supportsAndroid9(): Boolean {
        return SDK_INT >= Build.VERSION_CODES.P
    }

    /**
     * API level 29
     */
    fun supportAndroid10(): Boolean {
        return SDK_INT >= Build.VERSION_CODES.Q
    }

    /**
     * API level 31
     */
    fun supportsAndroid12(): Boolean {
        return SDK_INT >= Build.VERSION_CODES.S
    }

    private fun findSupportedAbis(): List<ABI> {
        val supportedAbis = Build.SUPPORTED_ABIS ?: return listOf()
        return supportedAbis.mapNotNull {
            when (it) {
                "arm64-v8a" -> ABI.ARM64_V8A
                "armeabi-v7a" -> ABI.ARMEABI_V7A
                "armeabi" -> ABI.ARMEABI
                "x86_64" -> ABI.X86_64
                "x86" -> ABI.X86
                "mips" -> ABI.MIPS
                "mips64" -> ABI.MIPS64
                else -> throw UnknownAbiException("unknown ABI '$it'")
            }
        }
    }

    class UnknownAbiException(message: String) : Exception(message)
}